module/zfs/zfs_vnops.c

```diff
@@ -4922,6 +4923,736 @@ convoff(struct inode *ip, flock64_t *lckdat, int  whence, offset_t offset)
+
+	if (BP_GET_LEVEL(bp) != 0)
+		return (0);
+
+	/*
+	 * Indirect block pointers must be remapped to reflect the real
+	 * physical offset and length.  The remapping is transparent to
+	 * the fiemap interface so no additional extent flags are set.
+	 */
+	spa_config_enter(spa, SCL_VDEV, FTAG, RW_READER);
+	if (spa_remap_blkptr(spa, &bp_copy, NULL, NULL))
+		bp = &bp_copy;
+	spa_config_exit(spa, SCL_VDEV, FTAG);
+
+	for (int i = 0; i < fm->fm_copies; i++) {
+		zfs_fiemap_entry_t *fe, *pfe;
+		avl_index_t idx;
+
+		/*
+		 * N.B. Embedded block pointers and holes are only added to
+		 * the fm_extents_trees[0], the additional trees are used
+		 * for redundant copies of data blocks.
+		 */
+		if (i > 0 && (BP_IS_HOLE(bp) || BP_IS_EMBEDDED(bp)))
+			continue;
+
+		fe = kmem_zalloc(sizeof (zfs_fiemap_entry_t), KM_SLEEP);
+		fe->fe_logical_start = zb->zb_blkid * fm->fm_block_size;
+
+		if (BP_IS_HOLE(bp)) {
+			fe->fe_logical_len = fm->fm_block_size;
+			fe->fe_flags |= FIEMAP_EXTENT_UNWRITTEN;
+		} else if (BP_IS_EMBEDDED(bp)) {
+			fe->fe_logical_len = BPE_GET_LSIZE(bp);
+			fe->fe_physical_start = 0;
+			fe->fe_physical_len = BPE_GET_PSIZE(bp);
+			fe->fe_flags |= FIEMAP_EXTENT_DATA_INLINE |
+			    FIEMAP_EXTENT_NOT_ALIGNED;
+
+			if (BP_IS_ENCRYPTED(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_DATA_ENCRYPTED;
+			if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF)
+				fe->fe_flags |= FIEMAP_EXTENT_ENCODED;
+		} else {
+			if (i >= BP_GET_NDVAS(bp)) {
+				kmem_free(fe, sizeof (zfs_fiemap_entry_t));
+				continue;
+			}
+
+			if (BP_IS_ENCRYPTED(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_DATA_ENCRYPTED;
+			if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF)
+				fe->fe_flags |= FIEMAP_EXTENT_ENCODED;
+			if (BP_GET_DEDUP(bp))
+				fe->fe_flags |= FIEMAP_EXTENT_SHARED;
+
+			/*
+			 * Report gang blocks as a single unknown extent.
+			 * Ideally we should be walking the gang block tree and
+			 * reporting all component-blocks as physical extents.
+			 */
+			if (BP_IS_GANG(bp)) {
+				fe->fe_flags |= FIEMAP_EXTENT_UNKNOWN;
+				fe->fe_physical_start = 0;
+				fe->fe_physical_len = 0;
+				fe->fe_vdev = 0;
+			} else {
+				fe->fe_physical_len = BP_GET_PSIZE(bp);
+
+				if (DVA_IS_VALID(&bp->blk_dva[i])) {
+					fe->fe_vdev =
+					    DVA_GET_VDEV(&bp->blk_dva[i]);
+					fe->fe_physical_start =
+					    DVA_GET_OFFSET(&bp->blk_dva[i]);
+				}
+			}
+
+			fe->fe_logical_len = BP_GET_LSIZE(bp);
+		}
+
+		/*
+		 * By default merge compatible adjacent block pointers in to a
+		 * single extent.  Embedded block pointers can never be merged.
+		 *
+		 * N.B. Block pointers provided by the iterator will always
+		 * be in logical offset order.  Therefore, it is sufficient
+		 * to check only the previously inserted entry when merging.
+		 */
+		pfe = avl_last(&fm->fm_extent_trees[i]);
+		if (pfe != NULL && !BP_IS_EMBEDDED(bp) &&
+		    !(fm->fm_flags & FIEMAP_FLAG_NOMERGE)) {
+			ASSERT3U(pfe->fe_logical_start + pfe->fe_logical_len,
+			    ==, fe->fe_logical_start);
```

Comment:
How is this ensured?  I don't think you're guaranteed to get a callback for every possible L0 block.  E.g. if an indirect block is a hole, I don't think we make calls for the (non-existent) L0's under it.

Reply:
That sounds right.  Let me investigate this and add a test for the case you're describing.  In which case we should be able to infer a hole due to the missing L0's and I'll go ahead and add or extent as needed.